File: Graph/csharp/PrereqsPossible.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class PrereqsPossible
{
    public static bool Solve(int numCourses, IEnumerable<(int, int)> prereqs)
    {
        var graph = new Dictionary<int, List<int>>();
        for (var i = 0; i < numCourses; i++)
        {
            graph[i] = new List<int>();
        }

        foreach (var (a, b) in prereqs)
        {
            graph[a].Add(b);
        }

        var visiting = new HashSet<int>();
        var visited = new HashSet<int>();

        for (var i = 0; i < numCourses; i++)
        {
            if (Detect(graph, i, visiting, visited))
            {
                return false;
            }
        }

        return true;
    }

    private static bool Detect(Dictionary<int, List<int>> graph, int node, HashSet<int> visiting, HashSet<int> visited)
    {
        if (visited.Contains(node))
        {
            return false;
        }
        if (visiting.Contains(node))
        {
            return true;
        }

        visiting.Add(node);
        foreach (var neighbor in graph[node])
        {
            if (Detect(graph, neighbor, visiting, visited))
            {
                return true;
            }
        }

        visiting.Remove(node);
        visited.Add(node);
        return false;
    }
}

File: Graph/php/prereqsPossible.php
<?php

declare(strict_types=1);

function prereqsPossible(int $numCourses, array $prereqs): bool
{
    $graph = [];
    for ($i = 0; $i < $numCourses; $i++) {
        $graph[$i] = [];
    }
    foreach ($prereqs as [$a, $b]) {
        $graph[$a][] = $b;
    }

    $visiting = [];
    $visited = [];

    for ($i = 0; $i < $numCourses; $i++) {
        if (prereqCycle($graph, $i, $visiting, $visited)) {
            return false;
        }
    }

    return true;
}

function prereqCycle(array $graph, int $node, array &$visiting, array &$visited): bool
{
    if (isset($visited[$node])) {
        return false;
    }
    if (isset($visiting[$node])) {
        return true;
    }

    $visiting[$node] = true;
    foreach ($graph[$node] as $neighbor) {
        if (prereqCycle($graph, (int) $neighbor, $visiting, $visited)) {
            return true;
        }
    }

    unset($visiting[$node]);
    $visited[$node] = true;
    return false;
}
